src/value.rs

```diff
@@ -113,6 +120,99 @@ impl<'gc> Value<'gc> {
         }
     }
 
+    // A small helper function to handle division-like zero handling
+    fn safe_div<T: PartialEq + ToPrimitive + Zero>(
+        lhs: T,
+        rhs: T,
+        f: &Fn(T, T) -> Value<'gc>,
+    ) -> Value<'gc> {
+        match (lhs, rhs) {
+            // Seems that all nans are negative in lua
+            (ref a, ref b) if a.is_zero() && b.is_zero() => Value::Number(-f64::NAN),
+            (ref a, ref b) if b.is_zero() => {
+                Value::Number(copysign(f64::INFINITY, a.to_f64().unwrap()))
+            }
+            (a, b) => f(a, b),
+        }
+    }
+
+    // This operation always returns a Number, even when called by int arguments
+    pub fn float_divide(self, other: Value<'gc>) -> Option<Value<'gc>> {
+        let (a, b) = match (self, other) {
+            (Value::Integer(a), Value::Integer(b)) => (a as f64, b as f64),
+            (Value::Number(a), Value::Number(b)) => (a, b),
+            (Value::Integer(a), Value::Number(b)) => (a as f64, b),
+            (Value::Number(a), Value::Integer(b)) => (a, b as f64),
+            _ => return None,
+        };
+
+        Some(Value::safe_div(a, b, &|a, b| Value::Number(a / b)))
+    }
+
+    pub fn floor_divide(self, other: Value<'gc>) -> Option<Value<'gc>> {
+        let (a, b) = match (self, other) {
+            // Seems that all nans are negative in lua
+            (Value::Integer(a), Value::Integer(b)) => {
+                return Some(Value::safe_div(a, b, &|a, b| {
+                    Value::Integer(a.wrapping_div(b))
+                }));
+            }
+            (Value::Number(a), Value::Number(b)) => (a, b),
+            (Value::Integer(a), Value::Number(b)) => (a as f64, b),
+            (Value::Number(a), Value::Integer(b)) => (a, b as f64),
+            _ => return None,
+        };
+
+        Some(Value::safe_div(a, b, &|a, b| {
+            Value::Number((a / b).floor())
+        }))
+    }
+
+    // When given a % b, lua computes the remainder, not the modulo.
+    // However, Rust computes the modulo correctly.  (e.g. -2 % 3 = 1 according to lua, and -2
+    // according to Rust.)
+    // This is why there is the second step.  Hopefully, the compiler will optimize the extra
+    // mod out
```

Comment:
I wonder if this behavior is dependent on the C compiler / architecture used to compile Lua, I bet it is.

Reply:
I'm hoping you're wrong, but I know you're right.

Reply:
Fun fact, Lua number parsing depends on the system locale.  This is why the target compatibility at the language level is "aggressive" not "complete".

Reply:
No, it doesn't depend on the C compiler, it's guaranteed to have C99 behavior: https://github.com/kyren/rlua/blob/e351bf57fa4c3985109d6674a8b6c35f9890da16/lua/lvm.c#L572

I don't *like* C99 behavior, but it is consistent.

Reply:
I'm glad to hear it's defined.  I'm noticing that you're attempting to avoid overflow in rlua, but the PUC Rio engine doesn't seem to even allow you to construct an Integer with values INT_MIN.

> -9223372036854775808
-9.2233720368548e+18
> -9223372036854775807
-9223372036854775807

Reply:
That file I linked you isn't rlua code, rlua embeds Lua 5.3.5, so I just linked you to the PUC-Rio Lua source inside it.

Or are you referring to rlua's behavior with `Value::Integer` not guarding against INT_MIN?

Reply:
Oh, of course.  You can't directly create a MIN_INT value, but one can arise as part of a calculation, so that check of course still makes sense.

\> -9223372036854775807 - 1  
-9223372036854775808  
\> (-9223372036854775807 - 1) * -1  
-9223372036854775808  
\> (-9223372036854775807 - 1) % 1  
0

Reply:
.....wat

Reply:
Well, okay now that I look at it it's not quite wat material, Lua allows you to observe all the weird integer overflow behavior as long as the literals you type all are integers, so that *kind of* makes sense... I haven't dug this deep into Lua's numeric behavior before.

Reply:
> I wonder if this behavior is dependent on the C compiler / architecture used to compile Lua, I bet it is.

No, see the Lua docs. Lua uses floor division, and modulus is defined similarly. This is better mathematically, as most repeating things don't have a discontinuity at 0!

Reply:
Okay, I see you found this. Sorry to pop up in your commit BTW. I must have 'watched' this!